cpp/src/ice2slice/Main.cpp

```diff
@@ -138,76 +138,56 @@ compile(const vector<string>& argv)
 
     ostringstream os;
 
-    // Create a copy of args without the duplicates.
-    vector<string> sources;
-    for (const auto& arg : args)
+    for (const auto& fileName : sliceFiles)
     {
-        auto p = find(sources.begin(), sources.end(), arg);
-        if (p == sources.end())
+        UnitPtr unit;
+        try
         {
-            sources.push_back(arg);
-        }
-    }
-
-    for (auto i = sources.begin(); i != sources.end();)
-    {
-        PreprocessorPtr icecpp = Preprocessor::create(argv[0], *i, cppArgs);
-        FILE* cppHandle = icecpp->preprocess(true, "-D__ICE2SLICE__");
-
-        if (cppHandle == nullptr)
-        {
-            return EXIT_FAILURE;
-        }
-
-        UnitPtr p = Unit::createUnit("", false);
-        int parseStatus = p->parse(*i, cppHandle, debug);
-
-        if (!icecpp->close())
-        {
-            p->destroy();
-            return EXIT_FAILURE;
-        }
+            PreprocessorPtr preprocessor = Preprocessor::create(argv[0], fileName, preprocessorArgs);
+            FILE* preprocessedHandle = preprocessor->preprocess("-D__ICE2SLICE__");
```

Comment:
Updated `preprocess` to throw an std::exception rather than returning null, to simplify error handling.